contracts/BonusRewards.sol

```diff
@@ -3,11 +3,363 @@
+    uint256 endTime = received * WEEK / bonus.weeklyRewards + bonus.endTime;
+
+    pools[_lpToken].bonuses[_poolBonusId].endTime = endTime;
+    pools[_lpToken].bonuses[_poolBonusId].remBonus = bonus.remBonus + received;
+  }
+
+  /// @notice add pools and authorizers to add bonus tokens for pools, combine two calls into one. Only reason we add pools is when bonus tokens will be added
+  function addPoolsAndAllowBonus(
+    address[] calldata _lpTokens,
+    address[] calldata _bonusTokenAddrs,
+    address[] calldata _authorizers
+  ) external override onlyOwner notPaused {
+    // add pools
+    for (uint256 i = 0; i < _lpTokens.length; i++) {
+      address _lpToken = _lpTokens[i];
+      Pool memory pool = pools[_lpToken];
+      if (pool.lastUpdatedAt == 0) {
+        pools[_lpToken].lastUpdatedAt = block.timestamp;
+        poolList.push(_lpToken);
+      }
+
+      // add bonus tokens and their authorizers (who are allowed to add the token to pool)
+      for (uint256 j = 0; j < _bonusTokenAddrs.length; j++) {
+        address _bonusTokenAddr = _bonusTokenAddrs[j];
+        require(pools[_bonusTokenAddr].lastUpdatedAt == 0, "BonusRewards: lpToken, not allowed");
+        allowedTokenAuthorizers[_lpToken][_bonusTokenAddr] = _authorizers;
+        if (bonusTokenAddrMap[_bonusTokenAddr] == 0) {
+          bonusTokenAddrMap[_bonusTokenAddr] = 1;
+        }
+      }
+    }
+  }
+
+  /// @notice collect bonus token dust to treasury
+  function collectDust(address _token, address _lpToken, uint256 _poolBonusId) external override onlyOwner {
+    require(pools[_token].lastUpdatedAt == 0, "BonusRewards: lpToken, not allowed");
+
+    if (_token == address(0)) { // token address(0) = ETH
+      payable(owner()).transfer(address(this).balance);
+    } else {
+      uint256 balance = IERC20(_token).balanceOf(address(this));
+      if (bonusTokenAddrMap[_token] == 1) {
+        // bonus token
+        Bonus memory bonus = pools[_lpToken].bonuses[_poolBonusId];
+        require(bonus.bonusTokenAddr == _token, "BonusRewards: wrong pool");
+        require(bonus.endTime + WEEK < block.timestamp, "BonusRewards: not ready");
+        balance = bonus.remBonus;
+        pools[_lpToken].bonuses[_poolBonusId].remBonus = 0;
+      }
+
+      IERC20(_token).transfer(owner(), balance);
+    }
+  }
+
+  function setResponders(address[] calldata _responders) external override onlyOwner {
+    responders = _responders;
+  }
+
+  function setPaused(bool _paused) external override {
+    require(_isAuthorized(msg.sender, responders), "BonusRewards: caller not responder");
+    paused = _paused;
+  }
+
+  /// @notice update pool's bonus per staked token till current block timestamp, do nothing if pool does not exist
+  function _updatePool(address _lpToken) private {
+    Pool storage pool = pools[_lpToken];
+    uint256 poolLastUpdatedAt = pool.lastUpdatedAt;
+    if (poolLastUpdatedAt == 0 || block.timestamp <= poolLastUpdatedAt) return;
```

Comment:
It should be impossible to get to this point with poolLastUpdatedAt  == 0.

Reply:
It is possible to be called on a non-exist pool by passing a random lpToken that does not have a pool in this contract.
 
I dont think we should revert.  `claimRewardsForPools` call this func multiple times, we should claim for the correct pools even if they passed some/one bad pools

Reply:
IMO I don't see a reason for us to not revert on invalid pools. If the transaction will be uneventful it should fail.

Reply:
So all single pool actions already check if pool exists (deposit, withdraw, etc.).

only one does not is `claimRewardsForPools`. IMO, if it is called, we should claim as much as possible for the user. Instead of trying to claim a bunch and then revert, costs lots of gas.

Reply:
On another thought, this is fine. Wouldn't want to risk locking a transaction somehow.